### PR TITLE
Use kube::release::gcs::publish_latest_official logic in mark-stable-release.sh

### LINKED ISSUE #14785: mark-stable-release.sh logic should be consolidated with kube::release::gcs::publish_latest_official
(cc @zmerlynn @brendandburns)

Despite doing very, very similar things, `mark-stable-release.sh` doesn't rely on `kube::release::gcs::publish_latest_official`.  It should, or the logic should otherwise be consolidated.

I'm going to maybe work on this to make #14771 easier.

### PR DESCRIPTION
**Pull request by ikehz**
Fixes #14785.

@zmerlynn please give this a very skeptical review.  I haven't checked to make sure this works properly; wanted to check with you about how to do so.  My intention was to replace `KUBE_GCS_RELEASE_BUCKET` with my own bucket and run through things.  Concerns about that approach?

### CODE REVIEW

**Comment by zmerlynn** [comment]
> My intention was to replace KUBE_GCS_RELEASE_BUCKET with my own bucket and run through things. Concerns about that approach?

No, I usually test these things by commenting out `KUBE_GCE_RELEASE_BUCKET`, which triggers the logic that puts it in the personal hash bucket. But you can also override it.

File: build/common.sh [hunk 1]
```diff
@@ -1007,19 +1007,47 @@ function kube::release::gcs::copy_release_artifacts() {
...
 
 function kube::release::gcs::publish_latest() {
-  local latest_file_dst="gs://${KUBE_GCS_RELEASE_BUCKET}/${KUBE_GCS_LATEST_FILE}"
+  local -r latest_file_dst="gs://${KUBE_GCS_RELEASE_BUCKET}/${KUBE_GCS_LATEST_FILE}"
 
   mkdir -p "${RELEASE_STAGE}/upload" || return 1
   echo ${KUBE_GCS_LATEST_CONTENTS} > "${RELEASE_STAGE}/upload/latest" || return 1
 
+  local -r release_dir="gs://${KUBE_GCS_RELEASE_BUCKET}/${KUBE_GCS_RELEASE_PREFIX}"
+  if ! gsutil ls ${release_dir}; then
```
Comment (zmerlynn) [inline build/common.sh]: Nit: Quote the variable: https://google-styleguide.googlecode.com/svn/trunk/shell.xml#Quoting

But also, this should redirect output. See how `kube::release::gcs::ensure_release_bucket` does it, for instance.

File: build/common.sh [hunk 2]
```diff
@@ -1007,19 +1007,47 @@ function kube::release::gcs::copy_release_artifacts() {
...
   if [[ ${KUBE_GCS_MAKE_PUBLIC} =~ ^[yY]$ ]]; then
     gsutil acl ch -R -g all:R "${latest_file_dst}" >/dev/null 2>&1 || return 1
+    # If public, validate public link
+    local -r public_link="https://storage.googleapis.com/${KUBE_GCS_RELEASE_BUCKET}/${KUBE_GCS_LATEST_FILE}"
+    kube::log::status "Validating uploaded version file at ${public_link}"
+    contents="$(curl -s ${public_link})"
+  else
+    # If not public, validate using gsutil
+    kube::log::status "Validating uploaded version file at ${latest_file_dst}"
+    contents="$(gsutil cat ${latest_file_dst})"
```
Comment (zmerlynn) [inline build/common.sh]: Quote `${latest_file_dst}`

File: build/common.sh [hunk 3]
```diff
@@ -1007,19 +1007,47 @@ function kube::release::gcs::copy_release_artifacts() {
...
+  fi
+}
 
-  kube::log::status "gsutil cat ${latest_file_dst}:"
-  gsutil cat ${latest_file_dst} || return 1
+# Validate a release version.
+# TODO(ihmccreery): as part of this refactor, the KUBE_GCS_LATEST_CONTENTS
+# should really be just a version.
+function kube::release::gcs::validate_latest_contents() {
+  local -r new_version=${KUBE_GCS_LATEST_CONTENTS}
```
Comment (zmerlynn) [inline build/common.sh]: quote

Comment (zmerlynn) [inline build/common.sh]: (even though I know you're just moving it)

File: build/mark-stable-release.sh [hunk 4]
```diff
@@ -20,31 +20,20 @@ set -o errexit
...
-if ! gsutil ls gs://kubernetes-release/release/${1}/kubernetes.tar.gz; then
-  echo "Release files don't exist, aborting."
-  exit 2
-fi
+KUBE_RELEASE_VERSION=${1-}
 
-STABLE_FILE_LOCATION="kubernetes-release/release/stable.txt"
+KUBE_GCS_MAKE_PUBLIC=y
+KUBE_GCS_RELEASE_BUCKET=kubernetes-release
+KUBE_GCS_RELEASE_PREFIX=release/${KUBE_RELEASE_VERSION}
```
Comment (zmerlynn) [inline build/mark-stable-release.sh]: Quote throughout here while we're at it.

**Comment by ihmccreery** [comment]
@zmerlynn PTAL.

Added a whole bunch of quoting, and renamed some functions & variables now that we use them for pushing `stable.txt`, not just `latest.txt`.

**Comment by ihmccreery** [comment]
(I also tested it.)

**Comment by ihmccreery** [comment]
@k8s-bot test this again please.

**Comment by ihmccreery** [comment]
@zmerlynn PTAL, or should I get another reviewer since you're oncall?

**Comment by zmerlynn** [comment]
Will look soon

**Comment by zmerlynn** [comment]
LGTM. @k8s-oncall, my advice is to handle merge this at your leisure and watch the build - the PR builder might have done the right thing, but I've had a few odd cases where `common.sh` changes broke things.

**Comment by zmerlynn** [comment]
(And to be clear, if the `kubernetes-build` job passes and seems to be uploading the right artifacts, you've won. Alternately, pass the merge torch to me for this one briefly and I'll handle it.)

**Comment by ihmccreery** [comment]
@zmerlynn Do you want me to squash commits?

**Comment by zmerlynn** [comment]
Please (I didn't notice)

**Comment by ihmccreery** [comment]
I'm closing this in favor of #14959, since it's diffbased against this, but the merge didn't go properly.  @zmerlynn, let me know if that's not okay, and we can reopen this.

**Pull request closed without merging**